nhost/service/manager.go

```diff
@@ -86,13 +95,162 @@ func (m *dockerComposeManager) SetGitBranch(gitBranch string) {
+		m.status.Error("Failed to list running services")
+		m.l.WithError(err).Debug("Failed to list running services")
+		return err
+	}
+
+	m.setProcessToStartInItsOwnProcessGroup(cmd)
+	out, err := cmd.CombinedOutput()
+	if err != nil && ctx.Err() != context.Canceled {
+		m.status.Error("Failed to list running services")
+		m.l.WithError(err).Debug("Failed to list running services")
+		return err
+	}
+
+	var svcs []composeSvc
+	err = json.Unmarshal(out, &svcs)
+	if err != nil {
+		m.status.Error("Failed to list running services")
+		m.l.WithError(err).Debug("Failed to list running services")
+		return err
+	}
+
+	runningSvcs := make(map[string]composeSvc)
+
+	for _, svc := range svcs {
+		runningSvcs[svc.Service] = svc
+	}
+
+	randomizerFunc := func(port uint32, timeout time.Duration) (uint32, error) {
+		t := time.After(timeout)
+
+		for i := 0; i < 1024; i++ {
+			select {
+			case <-t:
+				return 0, fmt.Errorf("timed out on ports randomize")
+			default:
+				if util.PortAvailable(fmt.Sprint(port)) {
+					return port, nil
+				}
+
+				port++
+			}
+		}
+
+		return 0, fmt.Errorf("failed to find available port")
+	}
+
+	portPublishedForSvc := func(port uint32, svc composeSvc) bool {
+		for _, p := range svc.Publishers {
+			if p.PublishedPort == 0 {
+				continue
+			}
+
+			if p.PublishedPort == port {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	for svcName, port := range m.ports {
+		if util.PortAvailable(fmt.Sprint(port)) {
+			continue
+		}
+
+		if svcName == compose.HasuraConsole || svcName == compose.HasuraConsoleApiPort {
+			randomPort, err := randomizerFunc(port, time.Second*10)
+			if err != nil {
+				m.status.Error(fmt.Sprintf("Failed to find available port for %s", svcName))
+				m.l.WithError(err).Debugf("Failed to find available port for %s", svcName)
+				return err
+			}
+
+			m.ports[svcName] = randomPort
+			continue
+		}
+
+		// if service is running
+		if svc, ok := runningSvcs[svcName]; ok {
+			// the port is used by the service
+			if portPublishedForSvc(port, svc) {
+				continue
+			}
+
+			// otherwise, we need to randomize the port
+			randomPort, err := randomizerFunc(port+1, time.Second*10)
+			if err != nil {
+				m.status.Error(fmt.Sprintf("Failed to find available port for %s", svc.Service))
+				m.l.WithError(err).Debugf("Failed to find available port for %s", svc.Service)
+				return err
+			}
+
+			// assign the random port to the service
+			m.ports[svcName] = randomPort
+			continue
+		}
+
+		randomPort, err := randomizerFunc(port+1, time.Second*10)
+		if err != nil {
+			m.status.Error(fmt.Sprintf("Failed to find available port for %s", svcName))
+			m.l.WithError(err).Debugf("Failed to find available port for %s", svcName)
+			return err
+		}
+
+		// assign the random port to the service
+		m.ports[svcName] = randomPort
+	}
+
+	// reinitialize the compose config with new ports
+	m.composeConfig = compose.NewConfig(m.nhostConfig, m.ports, m.env, m.branch, m.projectName)
```

Comment:
In general this method is going to be very hard to use and extremely confusing. We call a method `checkPorts` but this method doesn't really check ports, what it really does is introduce a bunch of side-effects. It modifies `m.ports[svcName]` and `m.composeConfig`.

I think you should rewrite this method ensuring there are no side-effects. And instead write a function that calls something like `findAndSetAvailablePorts(...)`

As a matter of fact, you'd save yourself a good deal of complexity if this was done as part of some initialization method for the container object itself. Something like:

```
func NewContainer(...) (*Container, error) {
     port, err := findPort(...)
     ...
     return &Container{....}
}
```

Assuming you have such type of object to start building from (sorry, not familiar with the cli).